backend/datasource/caching.go

```diff
@@ -0,0 +1,179 @@
+package datasource
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+var (
+	// ErrorNotCached is returned when a cached value was not found in the internal cache
+	ErrorNotCached = errors.New("no cached response found")
+)
+
+// A CachedResponse is the combination of a response and an expiration timestamp
+type CachedResponse struct {
+	ExpiresAt time.Time
+	Frames    data.Frames
+}
+
+func getCacheKey(duration time.Duration, req backend.DataQuery) (string, error) {
+	m := map[string]interface{}{
+		"query":    req.JSON,
+		"interval": req.Interval,
+		"type":     req.QueryType,
+		"time_range": backend.TimeRange{
+			To:   req.TimeRange.To.Round(duration),
```

Comment:
I think making request have cachable time ranges, and data source caching should be independent concerns. The main technical reason is that even without a caching data source, the service behind the data source could likely be hitting its cache if we align time request ranges ( https://docs.google.com/document/d/1YYvmT2fodC77sXaloazUfuzH2UHBHexJk6tF_Vr34_k / https://github.com/grafana/grafana/issues/5190)

Reply:
Yeah rounding the time range was because of the lack of something like https://github.com/grafana/grafana/issues/5190, which is still an open feature request. Should I leave it for now, and then come back to this whenever something like that feature request is implemented?